File: app/src/test/java/no/nav/tilbakemeldingsmottak/consumer/email/aad/EmailServiceTest.java
package no.nav.tilbakemeldingsmottak.consumer.email.aad;

import com.microsoft.graph.models.BodyType;
import com.microsoft.graph.models.Message;
import no.nav.tilbakemeldingsmottak.itest.ApplicationTest;
import org.junit.jupiter.api.Test;
import org.mockito.*;

import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class EmailServiceTest extends ApplicationTest {

    @Mock
    AADMailClient mailClient;

    @InjectMocks
    AzureEmailService emailService;

    @Captor
    ArgumentCaptor<Message> emailCaptor;

    @Test
    public void testSendMessage() throws Exception {
        String content = "This is a test message";
        String subject = "Test of messageService";
        String emailToAddress = "toMail@trygdeetaten.no";
        String emailToAddress2 = "extraReceiver@trygdeetaten.no";
        List to = Arrays.asList(emailToAddress, emailToAddress2);
        emailService.sendSimpleMessage(to, subject, content);

        Mockito.verify(mailClient).sendMailViaClient(emailCaptor.capture());
        Message value = emailCaptor.getValue();
        assertEquals(value.body.contentType, BodyType.HTML);
        assertEquals(emailToAddress, value.toRecipients.get(0).emailAddress.address);
        assertEquals(emailToAddress2, value.toRecipients.get(1).emailAddress.address);
        assertEquals(value.subject, subject);

    }

    @Test
    public void testSendMessageWithAttachment() throws Exception {
        String content = "This is a test message";
        String subject = "Test of messageService";
        String emailToAddress = "toMail@trygdeetaten.no";
        byte[] attachment = readFileFromClasspath("/litenPdf.pdf");
        String attachmentName = "serviceklage.pdf";

        emailService.sendMessageWithAttachments(emailToAddress, subject, content, attachment, attachmentName);

        Mockito.verify(mailClient).sendMailViaClient(emailCaptor.capture());
        Message value = emailCaptor.getValue();
        assertEquals(value.body.contentType, BodyType.HTML);
        assertEquals(value.toRecipients.get(0).emailAddress.address, emailToAddress);
        assertEquals(value.subject, subject);
        assertEquals(value.attachments.getCurrentPage().get(0).name, attachmentName);

    }

    private byte[] readFileFromClasspath(String filename) throws IOException {
        InputStream inputStream = EmailServiceTest.class.getResourceAsStream(filename);
        return inputStream.readAllBytes();
    }

}

File: app/src/test/kotlin/no/nav/tilbakemeldingsmottak/consumer/email/aad/EmailServiceTest.kt
package no.nav.tilbakemeldingsmottak.consumer.email.aad

import com.microsoft.graph.models.BodyType
import com.microsoft.graph.models.Message
import no.nav.tilbakemeldingsmottak.ApplicationTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.*
import java.io.IOException

internal class EmailServiceTest : ApplicationTest() {

    @Mock
    lateinit var mailClient: AADMailClient

    @InjectMocks
    lateinit var azureEmailService: AzureEmailService

    @Captor
    lateinit var emailCaptor: ArgumentCaptor<Message>

    @Test
    fun `test send message`() {
        // Given
        val content = "This is a test message"
        val subject = "Test of messageService"
        val emailToAddress = "toMail@trygdeetaten.no"
        val emailToAddress2 = "extraReceiver@trygdeetaten.no"
        val to = listOf(emailToAddress, emailToAddress2)

        // When
        azureEmailService.sendSimpleMessage(to, subject, content)

        // Then
        Mockito.verify(mailClient).sendMailViaClient(emailCaptor.capture())
        val value = emailCaptor.value
        assertEquals(BodyType.HTML, value.body?.contentType)
        assertEquals(emailToAddress, value.toRecipients?.get(0)?.emailAddress?.address)
        assertEquals(emailToAddress2, value.toRecipients?.get(1)?.emailAddress?.address)
        assertEquals(subject, value.subject)
    }

    @Test
    fun `test send message with attachment`() {
        // Given
        val content = "This is a test message"
        val subject = "Test of messageService"
        val emailToAddress = "toMail@trygdeetaten.no"
        val attachment = readFileFromClasspath("/litenPdf.pdf")
        val attachmentName = "serviceklage.pdf"

        // When
        azureEmailService.sendMessageWithAttachments(emailToAddress, subject, content, attachment, attachmentName)

        // Then
        Mockito.verify(mailClient).sendMailViaClient(emailCaptor.capture())
        val value = emailCaptor.value
        assertEquals(BodyType.HTML, value.body?.contentType)
        assertEquals(emailToAddress, value.toRecipients?.get(0)?.emailAddress?.address)
        assertEquals(subject, value.subject)
        assertEquals(attachmentName, value.attachments?.currentPage?.get(0)?.name)
    }

    @Throws(IOException::class)
    private fun readFileFromClasspath(filename: String): ByteArray {
        val inputStream = EmailServiceTest::class.java.getResourceAsStream(filename)
        return inputStream?.readAllBytes() ?: throw IOException("File not found")
    }
}
